### PR TITLE
Add division and modulo operator support

### PR DESCRIPTION
**Pull request by splasky**
This commit fixes the following known issues:
1.Division and modulo operator support from arm abi.
2.Add dlopen function call.
3.Update test case.

### CODE REVIEW

**Comment by splasky** [comment]
Import operator from abi may have following question:
1. Hard to port to another platform.
2. Not all arm cpu can support.
3. But this commit prove we can use other library.
Any ideas?

**Comment by jserv** [comment]
@lecopzer, can you help reviewing this pull request and relevant concerns?

File: amacc.c [hunk 1]
```diff
@@ -312,7 +312,7 @@ void next()
                              "SHL  SHR  ADD  SUB  MUL  "
                              "OPEN READ WRIT CLOS PRTF MALC FREE "
                              "MSET MCMP MCPY MMAP "
-                             "DSYM BSCH STRT EXIT CLCA" [*++le * 5]);
+                             "DSYM BSCH STRT DLOP EXIT DIV  MOD  CLCA " [*++le * 5]);
```
Comment (jserv) [inline amacc.c]: Should not extra blank right after `CLCA`.

File: amacc.c [hunk 2]
```diff
@@ -324,7 +324,7 @@ void next()
         case '\r':
             break;
         case '/':
-            if (*p == '/') { // comment
+	        if (*p == '/') { // comment
```
Comment (jserv) [inline amacc.c]: Don't change this line.

File: amacc.c [hunk 3]
```diff
@@ -336,8 +336,8 @@ void next()
                 }
                 ++p;
             } else {
-                // FIXME: Div is not supported
-                return;
+                if (*p == '=') { ++p; tk = DivAssign; }
+		        else tk = Div; return;
```
Comment (jserv) [inline amacc.c]: Indent.

File: amacc.c [hunk 4]
```diff
@@ -1720,11 +1756,12 @@ int elf32(int poolsz, int *main)
     append_strtab(&data, "");
     char *libc = append_strtab(&data, "libc.so.6");
     char *ldso = append_strtab(&data, "libdl.so.2");
+    char *ldccso = append_strtab(&data, "libgcc_s.so.1");
```
Comment (jserv) [inline amacc.c]: How about renaming it to `libgcc`?

Comment (splasky) [reply amacc.c]: GCC low level library named by libgcc_s.
Should we change it?

Comment (jserv) [reply amacc.c]: `libgcc_s` sounds good to me.

File: amacc.c (hunk 2 above)
Comment (splasky) [reply amacc.c]: I found this is the issue for my vscode.

File: amacc.c [hunk 5]
```diff
@@ -1231,6 +1250,23 @@ int *codegen(int *jitmem, int *jitmap)
...
+                if (!handle) fatal("libgcc_s.so.1 open error!");
+                tmp = (int) dlsym(handle, scnames[i - OPEN]);
+            }
+            *je++ = 0xe49d0004 | (1 << 12); // pop r1
+            *je++ = 0xe49d0004 | (0 << 12); // pop r0
+            *je++ = 0xe28fe000;                          // add lr, pc, #0
+            if (!imm0) imm0 = je;
+            *il++ = (int) je++ + 1;
+            *iv++ = tmp;
+            break;
```
Comment (lecopzer) [inline amacc.c]: No need to add these lines for DIV and MOD.
`default:` case shall handle all of the system calls.
It obviously you copy the code from `default:`, this should be fixed.

File: amacc.c [hunk 6]
```diff
@@ -1568,7 +1604,7 @@ int elf32(int poolsz, int *main)
      * (4 instruction * 4 bytes), so the first codegen and second codegen
      * have consistent code_size.
      */
-    int FUNC_NUM = EXIT - OPEN + 1;
+    int FUNC_NUM = MOD - OPEN + 1;
```
Comment (lecopzer) [inline amacc.c]: Don't touch EXIT, please make EXIT the last system call so you won't need to add bellow `case DIV` and `case MOD`

File: amacc.c [hunk 7]
```diff
@@ -1720,11 +1756,12 @@ int elf32(int poolsz, int *main)
     append_strtab(&data, "");
     char *libc = append_strtab(&data, "libc.so.6");
     char *ldso = append_strtab(&data, "libdl.so.2");
+    char *ldccso = append_strtab(&data, "libgcc_s.so.1");
 
-    int *func_entries = (int *) malloc(sizeof(int) * (EXIT + 1));
+    int *func_entries = (int *) malloc(sizeof(int) * (MOD + 1));
```
Comment (lecopzer) [inline amacc.c]: ditto

File: amacc.c [hunk 8]
```diff
@@ -1720,11 +1756,12 @@ int elf32(int poolsz, int *main)
...
     if (!func_entries) die("elf32: could not malloc func_entries table");
 
-    for (i = OPEN; i <= EXIT; i++)
+    for (i = OPEN; i <= MOD; i++)
```
Comment (lecopzer) [inline amacc.c]: ditto

File: amacc.c [hunk 9]
```diff
@@ -1736,7 +1773,7 @@ int elf32(int poolsz, int *main)
     memset(data, 0, SYM_ENT_SIZE);
     data += SYM_ENT_SIZE;
 
-    for (i = OPEN; i <= EXIT; i++)
+    for (i = OPEN; i <= MOD; i++)
```
Comment (lecopzer) [inline amacc.c]: ditto

File: amacc.c (hunk 5 above)
Comment (lecopzer) [inline amacc.c]: No need to add these lines for DIV and MOD.
`default:` case shall handle all of the system calls.
It obviously you copy the code from `default:`, this should be fixed.

Comment (splasky) [reply amacc.c]: The DIV and MOD is outside the `default` case for some reasons:
1. AST generated div and mod same as other operator, not like typical system calls.
2. JIT need to load `libgcc_s` in runtime.

Comment (splasky) [reply amacc.c]: Actually, I am confuse what is the function of line 1266 to line 1267?

Comment (lecopzer) [reply amacc.c]: I've tested on my local, `DIV` and `MOD` can be merge into default switch case.
You can leave this optimization in another PR, these lines is okay for me now.

Dear @jserv could you answer the line 1266 and 1267 for `il` and `iv`? 
I'm not familiar with the compiler part.

Comment (jserv) [reply amacc.c]: It was a historical kludge while I was to perform "MOV PC, LR" trick: (so-called Traditional ARM instruction)
http://www.toves.org/books/armsub/

That should be replaced by preferred instruction for ARMv7-A, `BX lr`.

Reference:
* https://wiki.ubuntu.com/ARM/Thumb2PortingHowto
* https://lists.denx.de/pipermail/u-boot/2017-March/282897.html

File: amacc.c [hunk 10]
```diff
@@ -1997,18 +2035,22 @@ int main(int argc, char **argv)
         "sizeof struct switch for while "
         "open read write close printf malloc free "
         "memset memcmp memcpy mmap "
-        "dlsym bsearch __libc_start_main exit __clear_cache void main";
+        "dlsym bsearch __libc_start_main dlopen __aeabi_idiv __aeabi_idivmod exit __clear_cache void main";
```
Comment (jserv) [inline amacc.c]: Split "dlopen __aeabi_idiv __aeabi_idivmod" into a new line.

File: amacc.c (hunk 5 above)
Comment (jserv) [reply amacc.c]: I tend to merge this pull request and then leave further refactoring and/or enhancements.

**Pull request merged**